stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -644,37 +644,80 @@ extension Unsafe${Mutable}RawBufferPointer {
+    _debugPrecondition(source.underestimatedCount * elementStride <= count,
       "insufficient space to accommodate source.underestimatedCount elements")
-    guard let base = baseAddress else {
+    guard let base = _position, let end = _end else {
       // this can be a precondition since only an invalid argument should be costly
-      _precondition(source.underestimatedCount == 0, 
+      _precondition(source.underestimatedCount == 0,
         "no memory available to initialize from source")
-      return (it, UnsafeMutableBufferPointer(start: nil, count: 0))
-    }  
+      return (unwritten, UnsafeMutableBufferPointer(start: nil, count: 0))
+    }
 
-    for p in stride(from: base, 
+    var written = 0
+    for p in stride(from: base,
       // only advance to as far as the last element that will fit
-      to: base + count - elementStride + 1, 
+      to: end - elementStride + 1,
       by: elementStride
     ) {
       // underflow is permitted -- e.g. a sequence into
       // the spare capacity of an Array buffer
-      guard let x = it.next() else { break }
+      guard let x = unwritten.next() else { break }
       p.initializeMemory(as: S.Element.self, repeating: x, count: 1)
-      formIndex(&idx, offsetBy: elementStride)
+      written += 1
+    }
+
+    let start = base.assumingMemoryBound(to: S.Element.self)
+    let buffer = UnsafeMutableBufferPointer(start: start, count: written)
+    return (unwritten, buffer)
+  }
+
+  /// Initializes the buffer's memory with the given elements, binding the
+  /// initialized memory to the elements' type.
+  ///
+  /// When calling the `initializeMemory(as:from:)` method on a buffer `b`,
+  /// the memory referenced by `b` must be uninitialized or initialized to a
+  /// trivial type, and must be properly aligned for accessing `C.Element`.
+  /// The buffer must contain sufficient memory to accommodate
+  /// `source.count`.
+  ///
+  /// This method initializes the buffer with elements from `source` until
+  /// `source` is exhausted. After calling `initializeMemory(as:from:)`,
+  /// the memory referenced by the returned `UnsafeMutableBufferPointer`
+  /// instance is bound and initialized to type `C.Element`.
+  ///
+  /// - Parameters:
+  ///   - type: The type of the elements to bind the buffer's memory to.
+  ///   - source: A collection of elements with which to initialize the buffer.
+  /// - Returns: An iterator to any elements of `source` that didn't fit in the
+  ///   buffer, and a typed buffer of the written elements. The returned
+  ///   buffer references memory starting at the same base address as this
+  ///   buffer.
+  @_alwaysEmitIntoClient
+  public func initializeMemory<C: Collection>(
+    as type: C.Element.Type, from source: C
+  ) -> (unwritten: C.Iterator, initialized: UnsafeMutableBufferPointer<C.Element>) {
+
+    let available = count / MemoryLayout<C.Element>.stride
+    _debugPrecondition(source.count <= available,
```

Comment:
Nit: If we keep the min call below, it may make sense to extract `source.count` to a local variable, instead of repeatedly accessing it in debug builds -- depending on what `C` is, the compiler may not be able to always optimize away the repeated calls, even if this is specialized.